### PR TITLE
Fixing race condition with EndpointSlice controller garbage collection

### LINKED ISSUE #91287: endpointslice creation and garbage collection are racy
**What happened**:

In the PMEM-CSI CI, we rapidly deploy and remove a custom resources which then owns all objects created for that resource. Deletion of those objects is handled through ownership references, i.e. deleting the CR is expected to cascade to all objects. This worked fine up to Kubernetes 1.17.

When starting to test with 1.18.2, we pretty reliably get into a state where the CR and the service objects owned by it are marked for deletion, but don't go away.

Debugging with the garbage collector's dependency graph (see https://kubernetes.slack.com/archives/C0EG7JC6T/p1589547091345700?thread_ts=1589546899.344300&cid=C0EG7JC6T) showed that endpointslices were blocking the deletion of the services and thus the CR. In the graph dump (= a snaphot in time), they were not even marked for deletion.

A closer look showed that endpointslices for the services were getting created and deleted continuously for over 15 minutes. Eventually, the garbage collector won and managed to delete endpointslices and services before new endpointslices were created.

**What you expected to happen**:

There seems to be a missing check in https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/endpointslice_controller.go: if the service is marked for deletion (i.e. has a deletionTimestamp), then the controller should not create new endpointslices.

**How to reproduce it (as minimally and precisely as possible)**:

I tried to boil it down to a script, but it just ran fine. The "easiest" way to reproduce it is to run PMEM-CSI's E2E testing, which is probably not easy enough :disappointed: 

**Anything else we need to know?**:

Disabling the EndpointSlice feature makes the problem go away.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.18.2
- Cloud provider or hardware configuration: QEMU virtual machines
- OS (e.g: `cat /etc/os-release`): Fedora
- Install tools: kubeadm

/sig network

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This ensures the EndpointSlice controller does not create EndpointSlices for Services that are being deleted. This also adds an e2e test to cover EndpointSlice garbage collection when the EndpointSlice controller is busy.

**Which issue(s) this PR fixes**:
Fixes #91287

**Special notes for your reviewer**:
Unfortunately I have still not been able to recreate the bug as described by #91287. I believe this should fix the root cause though. I will follow up on that bug for more information.

**Does this PR introduce a user-facing change?**:
```release-note
Fixing race condition with EndpointSlice controller garbage collection.
```

/sig network
/priority important-soon
/cc @pohly @freehan
/assign @liggitt

### CODE REVIEW

File: pkg/controller/endpointslice/endpointslice_controller.go [hunk 1]
```diff
@@ -313,6 +313,12 @@ func (c *Controller) syncService(key string) error {
 		return err
 	}
 
+	if service.DeletionTimestamp != nil {
+		// If a Service is being deleted, return immediately and leave any
+		// related EndpointSlices to be cleaned up by garbage collector.
+		return nil
```
Comment (liggitt) [inline pkg/controller/endpointslice/endpointslice_controller.go:319]: should anything with triggerTimeTracker and reconciler be done here similar to the NotFound case?

Comment (robscott) [reply pkg/controller/endpointslice/endpointslice_controller.go:322]: Updated, thanks!

File: test/e2e/network/endpointslice.go [hunk 2]
```diff
@@ -299,6 +300,113 @@ var _ = SIGDescribe("EndpointSlice", func() {
...
+
+		go func() {
+			defer wg.Done()
+			createTestPods(podClient, 200, 251, sharedLabels)
+		}()
+
+		go func() {
+			defer wg.Done()
+			deleteTestPods(podClient, 0, 100, sharedLabels)
+		}()
```
Comment (robscott) [inline test/e2e/network/endpointslice.go:374]: @liggitt Would the scale of this test be considered too disruptive? It's currently creating a total of 202 pods though less than 200 should exist at any one time. I was trying to find a way to simulate the EndpointSlice controller being busy without actually overloading the cluster with resources. This also has the benefit of exercising multiple EndpointSlices, something we'd need >100 pods for.

Comment (liggitt) [reply test/e2e/network/endpointslice.go:374]: hmm… our e2e struggles enough that I'd hesitate to intentionally add scaleish tests here. @wojtek-t , any thoughts?

File: pkg/controller/endpointslice/endpointslice_controller.go [hunk 3]
```diff
@@ -313,6 +313,15 @@ func (c *Controller) syncService(key string) error {
 		return err
 	}
 
+	if service.DeletionTimestamp != nil {
```
Comment (liggitt) [inline pkg/controller/endpointslice/endpointslice_controller.go:316]: do we need a similar fix for the endpoints controller?

Comment (abhat) [reply pkg/controller/endpointslice/endpointslice_controller.go:316]: I am fairly new to this code, but it looks like in the case of endpoints controller we explicitly delete the endpoint instead of gc to kick in for the NotFound case. The check for DeletionTimestamp is nevertheless missing so we may end up creating new endpoints potentially similar to the original issue reported.

Comment (abhat) [reply pkg/controller/endpointslice/endpointslice_controller.go:316]: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L372

Comment (robscott) [reply pkg/controller/endpointslice/endpointslice_controller.go:316]: The Endpoints controller actually does not set an owner reference, so there's not really a parallel here. I think the Endpoins controller might predate owner references. There's an [issue from March 2015](https://github.com/kubernetes/kubernetes/issues/6877) that notes that Endpoints were never cleaned up by the Endpoints controller. This was [fixed in April 2015 by a PR that added a cleanup function](https://github.com/kubernetes/kubernetes/pull/7821) on controller startup. That [checkLeftoverEndpoints() function continues to exist](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L549) in a mostly unchanged state. 

That cleanup function does make an important exception for Endpoints with the `control-plane.alpha.kubernetes.io/leader` annotation though, not quite sure if Endpoints used for leader election ever have a corresponding Service though. Assuming they don't, this would be a non-issue for garbage collection.

So to summarize:
1. The Endpoints controller does not use owner references and will not conflict with garbage collection.
2. @abhat's link above shows that Endpoints are cleaned up within the syncService() function when a Service can not be found. 
3. We could potentially extend that cleanup logic to include a Service with a deletionTimestamp, but I'm not sure that would make a meaningful difference or be worth the effort.
4. It may be worth updating the Endpoints controller to use owner references.

File: test/e2e/network/endpointslice.go (hunk 2 above)
Comment (wojtek-t) [reply test/e2e/network/endpointslice.go:374]: Yeah... Note that we're runnin 30 tests in parallel, and clusters we test it on have 5? 6? nodes. So it's pretty significant for them.

I'm not particularly afraid of running this test in large clusters, but for presubmit it may cause problems for random tests, which would make debugging it really hard.

File: pkg/controller/endpointslice/endpointslice_controller.go (hunk 3 above)
Comment (liggitt) [reply pkg/controller/endpointslice/endpointslice_controller.go:316]: > The Endpoints controller actually does not set an owner reference, so there's not really a parallel here.

Ah, right.

File: test/e2e/network/endpointslice.go (hunk 2 above)
Comment (liggitt) [reply test/e2e/network/endpointslice.go:374]: let's drop the load test from this and lean on the unit test.

File: pkg/controller/endpointslice/endpointslice_controller.go (hunk 3 above)
Comment (liggitt) [inline pkg/controller/endpointslice/endpointslice_controller.go:316]: I'm trying to think through what this means if the service has a custom finalizer and deletionTimestamp, and the custom finalizer takes noticeable time to remove. Returning here means any existing endpointslices will go stale, could contain addresses that don't exist any more, won't pick up new addresses.

What does kube-proxy do with services with a deletionTimestamp?

File: test/e2e/network/endpointslice.go (hunk 2 above)
Comment (robscott) [reply test/e2e/network/endpointslice.go:374]: I talked with @wojtek-t about this and it sounds like it could be a good use of the `[Serial]` tag. I've added that for this test. I think it still provides significant value as the only e2e test to exercise multiple Endpoint Slices.

File: pkg/controller/endpointslice/endpointslice_controller.go (hunk 3 above)
Comment (robscott) [reply pkg/controller/endpointslice/endpointslice_controller.go:316]: That's a great question. As far as I can tell, kube-proxy does not have any special treatment for Services with a deletionTimestamp specified. I think once a Service has begun the deletion process the primary goal is to close out existing connections, something that this would still enable. I can't think of a great way to not potentially conflict with garbage collection while also preventing stale EndpointSlices during the deletion process. 

@freehan do you have any ideas about what should be done during Service deletion? Trying to understand what we should do for EndpointSlices when Services with finalizers are being deleted. Is it acceptable to not change EndpointSlices at all once the deletion process begins?

Comment (freehan) [reply pkg/controller/endpointslice/endpointslice_controller.go:316]: Might be worth changing kube-proxy to ignore service with DeletionTimestamp. 

I think it is totally okay for endpointSlice Controller to stop managing EndpointSlices of Service with DeletionTimestamp. And leave it to K8s GarbageCollector to remove EndpointSlice objects once done.

File: test/e2e/network/endpointslice.go [hunk 4]
```diff
@@ -299,6 +300,113 @@ var _ = SIGDescribe("EndpointSlice", func() {
...
+				if len(esList.Items) < 2 {
+					framework.Logf("Waiting for second EndpointSlice to be created for %s/%s Service", f.Namespace.Name, svc.Name)
+					return false, nil
+				}
+
+				return true, nil
+			})
+			framework.ExpectNoError(err, "Timed out waiting for second EndpointSlice to be created for %s/%s Service", f.Namespace.Name, svc.Name)
+
+			err = cs.CoreV1().Services(f.Namespace.Name).Delete(context.TODO(), svc.Name, metav1.DeleteOptions{})
```
Comment (pohly) [inline test/e2e/network/endpointslice.go:391]: We use `metav1.DeletePropagationForeground` as `PropagationPolicy`. If that isn't the default(*) , the service is going to be deleted although there are still some endpointslices owned by it, right?

(*) If I remember (and understood) correctly, that depends on how the ownership reference is set up in the endpointslices. I don't have a cluster running right now, so I can't check that.

File: test/e2e/network/endpointslice.go [hunk 5]
```diff
@@ -299,6 +300,113 @@ var _ = SIGDescribe("EndpointSlice", func() {
...
+			framework.ExpectNoError(err, "Error deleting %s/%s Service", f.Namespace.Name, svc.Name)
+		}()
+
+		wg.Wait()
+
+		// Wait for EndpointSlices to be garbage collected.
```
Comment (pohly) [inline test/e2e/network/endpointslice.go:397]: When using `metav1.DeletePropagationForeground`, a simpler test is to wait for the service to be removed after marking it for deletion.

File: test/e2e/network/endpointslice.go (hunk 4 above)
Comment (pohly) [reply test/e2e/network/endpointslice.go:391]: endpointslices have `blockOwnerDeletion: true`. So does this block deletion of the service, regardless of the `PropagationPolicy`? https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion describes `blockOwnerDeletion` but doesn't say anything about how it is handled when foreground cascading deletion is not used.

That then begs the question whether kubectl used foreground cascading deletion in the test script that I was trying to write. I am not sure anymore because in an experiment with `kubectl delete deployment/service-stress-test`, the deployment was deleted although there was a service which had it as owner with `blockOwnerDeletion: true` and the service itself couldn't be deleted because I had set a finalizer.

I was expecting `kubectl` to use foreground cascading based on https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#setting-the-cascading-deletion-policy, specifically "kubectl also supports cascading deletion. ... The default value for --cascade is true."

But I misunderstood `--cascading`: `kubectl -v=10` shows that `--cascade` only switches between `propagationPolicy":"Orphan"` and `"propagationPolicy":"Background"`. I don't see an option to select `propagationPolicy":"Foreground"`.

After switching to `curl` with `"propagationPolicy":"Foreground"`, I managed to trigger the race condition in my test script. I've posted it in the issue: https://github.com/kubernetes/kubernetes/issues/91287#issuecomment-633494000

File: pkg/controller/endpointslice/endpointslice_controller.go (hunk 3 above)
Comment (liggitt) [reply pkg/controller/endpointslice/endpointslice_controller.go:316]: > I can't think of a great way to not potentially conflict with garbage collection while also preventing stale EndpointSlices during the deletion process.

An alternative is to modify `reconciler#finalize` to skip creations if the service's deletionTimestamp is set, but to permit updates/deletes. That ensures existing endpointslices stay updated and that we prune stale endpoints, but that we don't create new EndpointSlice objects for a service being deleted.

Comment (robscott) [reply pkg/controller/endpointslice/endpointslice_controller.go]: Good idea - I've switched to that approach.

File: test/e2e/network/endpointslice.go (hunk 4 above)
Comment (robscott) [reply test/e2e/network/endpointslice.go]: Thanks for the detailed steps to reproduce this issue! I've added an e2e test that uses a similar approach to this. Although that script will still appear to fail for me on an e2e cluster with this fix, it is because the deployment sometimes sticks around for an extra second or two. In my testing, EndpointSlices are consistently empty in the script output with these fixes.

There is unfortunately no way to completely guarantee that the EndpointSlice controller will not try to create EndpointSlices after Service deletion begins, but this should prevent it from happening as soon as the EndpointSlice controller is aware the Service is being deleted.

File: test/e2e/network/endpointslice.go (hunk 5 above)
Comment (robscott) [reply test/e2e/network/endpointslice.go:524]: Thanks! I've added an e2e test that uses this approach.

File: pkg/controller/endpointslice/endpointslice_controller.go (hunk 3 above)
Comment (robscott) [reply pkg/controller/endpointslice/endpointslice_controller.go]: @freehan I am a bit torn on the best approach here. In some sense, updating kube-proxy to consider Services with a deletionTimestamp deleted/gone would result in the most consistent user experience. On the other hand, that does feel relatively harsh, especially if there are finalizers that rely on connectivity continuing for some period of time. Although this will result in a potentially inconsistent experience, it seems like it could be better and easier to understand than dropping everything immediately. Admittedly I'm not sure how much people care about connections on Services that are being deleted.

File: pkg/controller/endpointslice/endpointslice_controller_test.go [hunk 6]
```diff
@@ -101,6 +101,25 @@ func TestSyncServiceNoSelector(t *testing.T) {
 	assert.Len(t, client.Actions(), 0)
 }
 
+// Ensure SyncService for service with pending deletion results in no action
```
Comment (liggitt) [inline pkg/controller/endpointslice/endpointslice_controller_test.go:104]: is it worth separately testing that update/delete still occurs?

File: test/e2e/network/endpointslice.go [hunk 7]
```diff
@@ -299,6 +303,237 @@ var _ = SIGDescribe("EndpointSlice", func() {
...
+					Name:       "example",
+					Port:       80,
+					TargetPort: intstr.FromInt(3000),
+					Protocol:   v1.ProtocolTCP,
+				}},
+			},
+		})
+
+		lSelector := discoveryv1beta1.LabelServiceName + "=" + svc.Name
+		esWatch, err := cs.DiscoveryV1beta1().EndpointSlices(f.Namespace.Name).Watch(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
```
Comment (liggitt) [inline test/e2e/network/endpointslice.go:362]: this is starting a watch with no resourceVersion, which means it gets a synthetic add event for the existing items... is that what you wanted?

File: test/e2e/network/endpointslice.go [hunk 8]
```diff
@@ -299,6 +303,237 @@ var _ = SIGDescribe("EndpointSlice", func() {
...
+
+		deletePropagationForeground := metav1.DeletePropagationForeground
+		err = cs.AppsV1().Deployments(f.Namespace.Name).Delete(context.TODO(), deployment.Name, metav1.DeleteOptions{
+			PropagationPolicy: &deletePropagationForeground,
+		})
+		framework.ExpectNoError(err, "Error deleting %s/%s Deployment", f.Namespace.Name, deployment.Name)
+
+		// There's some delay between when a Service is deleted and the
+		// EndpointSlice informer is updated with that information. This
+		// allows for up to 1 second of delay.
```
Comment (liggitt) [inline test/e2e/network/endpointslice.go:374]: this seems likely to be flaky... is this trying to compensate for the fact that GC can observe and delete the endpointslice, and the endpointslicecontroller can observe that deletion and react to it before it observes the service deletion?

Comment (liggitt) [reply test/e2e/network/endpointslice.go:374]: would it be simpler to place an upper bound on the number of recreations (maybe 2?) and amount of time we'll accept (maybe 5 seconds) before we expect endpointslices to stop being recreated, and a time limit (maybe a minute?) for the service and all endpointslices to be gone?

Comment (robscott) [reply test/e2e/network/endpointslice.go]: That makes sense. I've updated this test to use that kind of approach.

File: test/e2e/network/endpointslice.go (hunk 7 above)
Comment (robscott) [reply test/e2e/network/endpointslice.go]: Good catch, thanks! Should be fixed now.

File: pkg/controller/endpointslice/endpointslice_controller_test.go (hunk 6 above)
Comment (robscott) [reply pkg/controller/endpointslice/endpointslice_controller_test.go:104]: I've added test coverage for this in `reconciler_test.go`. It's actually a relatively complex thing to test since the combination of creates and deletes will be converted into updates, so testing there was slightly more straightforward.

File: test/e2e/network/endpointslice.go [hunk 9]
```diff
@@ -299,6 +303,180 @@ var _ = SIGDescribe("EndpointSlice", func() {
...
+					Name:       "example",
+					Port:       80,
+					TargetPort: intstr.FromInt(3000),
+					Protocol:   v1.ProtocolTCP,
+				}},
+			},
+		})
+
+		// Create an initial set of 100 test pods for initial EndpointSlice.
+		createTestPods(podClient, 0, 100, sharedLabels)
```
Comment (liggitt) [inline test/e2e/network/endpointslice.go:404]: are we sure we'll always have 100 pods' worth of capacity here? @smarterclayton, I have vague memories of you writing an e2e that tried to scale down to cluster capacity at small sizes... do you remember that?

**Comment by liggitt** [comment]
The controller change and unit tests lgtm. I didn't review the e2e super-closely; I had one question about overall cluster capacity on the serial 100-pod test

File: test/e2e/network/endpointslice.go (hunk 9 above)
Comment (smarterclayton) [reply test/e2e/network/endpointslice.go:404]: Certainly we have enforced in tests that we have to be able to test on just a few small nodes (like 2, with maybe 1-2 cores each).  I *think* we already have tests at 100 pods in serial, but maybe not in parallel.

Comment (robscott) [reply test/e2e/network/endpointslice.go:404]: Looks like there's at least 1 serial e2e test running 100 pods _per node_. (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/node/kubelet_perf.go#L254 and https://github.com/kubernetes/kubernetes/blob/master/test/e2e/node/kubelet_perf.go#L69).

Comment (liggitt) [reply test/e2e/network/endpointslice.go:404]: that also has a Feature tag (`Feature:RegularResourceUsageTracking`) so I don't think it runs in default test configurations

Comment (robscott) [reply test/e2e/network/endpointslice.go:404]: Good catch, I can remove this test if it's an issue. Or maybe I could add a `[Feature:EndpointSlice]` tag here to have a similar limitation. It would be nice if we could have some kind of e2e test coverage of an EndpointSlice being filled up and overflowing into another one, but maybe we just need to rely on unit tests for that? It seems like the only alternative would be a separate set of e2e tests that ran the controller with less endpoints per slice.

Comment (liggitt) [reply test/e2e/network/endpointslice.go:404]: if we're wanting to backport this fix (and I think we should), I would probably drop this e2e. exercising the controller in an integration test with a lower endpoints-per-slice setting with a simulated set of ready pods controlled by the test might be more sensible... you could exercise the apiserver/gc/endpoints interactions without actually starting 100 processes on a remote node somewhere

Comment (robscott) [reply test/e2e/network/endpointslice.go]: Good call, an integration test does seem ideal here. As far as the e2e test, I tried one more approach that I think might still be helpful here. Instead of using lots of pods to get multiple EndpointSlices, I used named ports with unique port numbers on each pod. That means that each Pod will get its own EndpointSlice. The test now uses a total of 6 Pods and ends up with as many as 6 EndpointSlices (depending on the timing of the Service deletion). I think it's still potentially helpful to have test coverage ensuring that Service deletion can still work when the EndpointSlice controller would otherwise actively be trying to create EndpointSlices, and maybe more importantly test coverage ensuring that multiple EndpointSlices can be reliably garbage collected. With all that said, I'm still open to removing this entirely if it would help the change be cherry picked back.

Comment (liggitt) [reply test/e2e/network/endpointslice.go]: My recommendation generally is to test at the lowest level where a flawed interaction can be reproduced. At first glance, that would seem to be in an integration test (though I'm not sure how hard mocking the "ready" pods driving the endpointslice controller would be). If that proves difficult, the slimmed down e2e that is 1:1 pod:endpoint slice sounds like a good backup plan.

File: test/e2e/network/endpointslice.go [hunk 10]
```diff
@@ -299,6 +303,146 @@ var _ = SIGDescribe("EndpointSlice", func() {
...
+				Replicas: utilpointer.Int32Ptr(1),
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: sharedLabels,
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:  "container1",
+								Image: imageutils.GetE2EImage(imageutils.Nginx),
```
Comment (aojea) [inline test/e2e/network/endpointslice.go:329]: nginx was not ipv6 friendly (it does not listen in IPv6 by default)
, but if we don't have to poll the port, should not be better using busybox or a lighter image?

File: test/e2e/network/endpointslice.go [hunk 11]
```diff
@@ -299,6 +303,146 @@ var _ = SIGDescribe("EndpointSlice", func() {
...
+
+		deletePropagationForeground := metav1.DeletePropagationForeground
+		err = cs.AppsV1().Deployments(f.Namespace.Name).Delete(context.TODO(), deployment.Name, metav1.DeleteOptions{
+			PropagationPolicy: &deletePropagationForeground,
+		})
+		framework.ExpectNoError(err, "Error deleting %s/%s Deployment", f.Namespace.Name, deployment.Name)
+
+		// There's some delay between when a Service is deleted and the
+		// EndpointSlice informer is updated with that information. This
+		// allows for up to 2 EndpointSlice creations within 15 seconds of
```
Comment (aojea) [inline test/e2e/network/endpointslice.go:372]: is this time deterministic?
I'm concerned about possible flakes because of being too optimistic with the timings :sweat_smile:

File: test/e2e/network/endpointslice.go (hunk 10 above)
Comment (liggitt) [reply test/e2e/network/endpointslice.go:329]: yeah, just a pause image seems workable if we're not actually exercising the network

Comment (robscott) [reply test/e2e/network/endpointslice.go]: Updated, thanks!

**Comment by robscott** [comment]
Thanks to everyone for the great reviews on this! It does not seem like my initial more complex e2e tests were a good idea here, so I've removed them. Although an integration test might be helpful here, I haven't found a great way to properly test not conflicting with garbage collection without some form of an e2e test. So with that said, I've left in a simplified e2e test that should cover this potential conflict. Open to ideas on how I can improve this further. Also open to removing the e2e test entirely and just relying on the unit test.

File: test/e2e/network/endpointslice.go (hunk 11 above)
Comment (robscott) [reply test/e2e/network/endpointslice.go]: I made this even more lenient so it should never fail, and if it did, I think it would represent a real issue. With that said, I'm open to removing it altogether.

**Comment by aojea** [comment]
> Also open to removing the e2e test entirely and just relying on the unit test.

I vote for dropping the e2e test because I don't know if this assumption can be always true, and this is a very complex test for troubleshooting flakiness, at least for me :sweat_smile: 

> 		// There's some delay between when a Service is deleted and the
>		// EndpointSlice informer is updated with that information. This
>		// allows for up to 5 EndpointSlice creations within 15 seconds of
>		// deletion.

question, is not the e2e framework exercising this scenario?
once a test fails it deletes the namespace and its objects, maybe we can rely on that?

**Comment by pohly** [comment]
> question, is not the e2e framework exercising this scenario?
> once a test fails it deletes the namespace and its objects, maybe we can rely on that?

The problem only occurs when using foreground deletion, something that the E2E framework (most likely) doesn't do.

**Comment by liggitt** [comment]
I'm actually pretty satisfied with the unit test coverage... it proves the controller will never try to create an endpointslice for a service that is being deleted

**Comment by robscott** [comment]
Thanks for the great feedback! I've gone ahead and dropped the e2e test for this.

**Comment by runzexia** [comment]
could we cherrypick this to release v1.18 ...
